src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -101,3 +118,152 @@ func (r *downloadRequest) Validate() *util.JSONResponse {
 	}
 	return nil
 }
+
+func (r *downloadRequest) doDownload(w http.ResponseWriter, cfg *config.MediaAPI, db *storage.Database, activeRemoteRequests *types.ActiveRemoteRequests) *util.JSONResponse {
+	// check if we have a record of the media in our database
+	mediaMetadata, err := r.getMediaMetadata(db)
+	if err == nil {
+		// If we have a record, we can respond from the local file
+		r.MediaMetadata = mediaMetadata
+		return r.respondFromLocalFile(w, cfg.AbsBasePath)
+	} else if err == errDBNotFound {
+		if r.MediaMetadata.Origin == cfg.ServerName {
+			// If we do not have a record and the origin is local, the file is not found
+			r.Logger.WithError(err).Warn("Failed to look up file in database")
+			return &util.JSONResponse{
+				Code: 404,
+				JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+			}
+		}
+		// TODO: If we do not have a record and the origin is remote, we need to fetch it and respond with that file
+		return &util.JSONResponse{
+			Code: 404,
+			JSON: jsonerror.NotFound(fmt.Sprintf("File with media ID %q does not exist", r.MediaMetadata.MediaID)),
+		}
+	}
+	// Another error from the database
+	r.Logger.WithError(err).WithFields(log.Fields{
+		"MediaID": r.MediaMetadata.MediaID,
+		"Origin":  r.MediaMetadata.Origin,
+	}).Error("Error querying the database.")
+	return &util.JSONResponse{
+		Code: 500,
+		JSON: jsonerror.InternalServerError(),
+	}
+}
+
+// getMediaMetadata queries the database for media metadata
+func (r *downloadRequest) getMediaMetadata(db *storage.Database) (*types.MediaMetadata, error) {
+	mediaMetadata, err := db.GetMediaMetadata(r.MediaMetadata.MediaID, r.MediaMetadata.Origin)
+	if err != nil {
+		if err == sql.ErrNoRows {
```

Comment:
Having done that you can probably get rid of this function and the errDBNotFound and the errDBQuery variables.